### ISSUE TITLE
Calling Horovod as a library

### ISSUE THREAD

**Issue by jmjwozniak** [uber/horovod#194]
I am trying to call Horovod as a library from a parent MPI C program.  I made the following modification:

https://github.com/jmjwozniak/horovod/commit/b5e7d7c4ff0c37348c7b9bbd47528cbf33a51d59

to eliminate MPI_COMM_WORLD but there is a problem at shut down where the communicator is destroyed but the BackgroundThreadLoop is still running.  I would like to be able to restart Horovod on the same resources after shut down.   

My C interface is here: 

https://github.com/jmjwozniak/horovod/blob/master/controller/controller.h

Do you have any tips on doing this?

**Comment by jmjwozniak** [comment]
My most critical question: Is there any state in Horovod or TensorFlow that cannot be reinitialized for a second invocation as part of the same process?

**Comment by alsrgv** [comment]
@jmjwozniak, I think it should be fine to shut down and re-initialize Horovod.

You could guard `MPI_Init_thread()` with `!MPI_Initialized()` and move `MPI_Finalize()` into `HorovodGlobalState` destructor and guard with `!MPI_Finalized()`.

Then you could add a function that would stop Horovod along these lines:
```
    if (state.background_thread.joinable()) {
      state.shut_down = true;
      state.background_thread.join();
      state.initialization_done = false;
    }
```

If you find this very useful, we could take a PR with such changes.

**Comment by jmjwozniak** [comment]
Ah- I may have fixed it with this change at the end of the run:   
```
  for (auto it = state.tensor_fusion_buffers.begin();
       it != state.tensor_fusion_buffers.end(); it++) {
    // delete it->second;
    it->second = nullptr;
  }
```

**Comment by alsrgv** [comment]
@jmjwozniak, I think this PR, once completed, will help you - https://github.com/uber/horovod/pull/218.  You'll be able to do `horovod_finalize()` followed by `horovod_init()`.

What's the purpose of having a custom communicator for you?  Would https://github.com/uber/horovod/issues/105#issuecomment-373977819 solve your need?

**Comment by jmjwozniak** [comment]
This basically works as of: https://github.com/jmjwozniak/horovod/commit/ea715e13c23931fcca63997ed4de3be888838522

**Comment by jmjwozniak** [comment]
The next issue is a nicer interface to pass the communicator to Horovod from the parent

**Comment by jmjwozniak** [comment]
Yes, #218 is related.  However, the goal of the custom communicators is to run multiple Horovod instances inside a big MPI job that is doing, e.g., hyperparameter optimization.  The changes and tests in my fork handle MPI_Init/Finalize, allow the Python user to stop the background thread, and re-initialize the tensor buffers for the next Horovod job within the same resources.

**Comment by alsrgv** [comment]
@jmjwozniak, how big is your big MPI job?  Would doing it with #105 work, i.e. just specify ranks via `hvd.init(default_ranks=[1,3,5,...])`?  I'm wondering if the solution for per-operation ranks could be useful to you, too.  Caveat: `hvd.rank()` would still return `COMM_WORLD` ranks.

**Comment by jmjwozniak** [comment]
Pretty big- we are using large Crays to do these runs as described here: http://www.mcs.anl.gov/~wozniak/papers/CANDLE_2017.pdf

**Comment by alsrgv** [comment]
OK, so it sounds like there's a room for two features.

One feature is to create a small horovod in a "sandbox" of few ranks, these sandboxes are totally independent.  This can be implemented by creating a group with list of ranks and then creating a communicator from that group.  These sandboxed Horovods can come and go over a lifetime of an MPI application.  E.g. hyperparameter tuning.

Second feature defined in #105 would allow to have per-operation groups within a sandbox, but all such groups would have the same lifetime.  E.g. Allreduce gradients over all ranks and Allreduce batch norm on every 4 ranks.

We'd be happy to take PRs for both.  After #218 is merged, first feature should be a small change.

**Comment by ajbouh** [comment]
Any update here?

Would love to be able to use Horovod as a library embedded in a larger application with a lifecycle that's not managed by mpirun!

**Comment by alsrgv** [comment]
@ajbouh, @lgarithm (since you upvoted) - can you share your use cases?  Would you need programmatic (re)initialization/termination (#218), allreduce on static subgroups (partition Horovod communicator during initialization), or allreduce on dynamic subgroups (specify set of ranks per operation)?

**Comment by ajbouh** [comment]
My use case is essentially: run Horovod from the Python REPL using a *pre-allocated* set of workers.

To do this I need to do two things differently than Horovod documentation suggests:

- start the proper number of worker machines and connect them to one another *without* knowing what graph they'll be running.
- transmit the fragments of training graph needed by each worker dynamically. (This is similar to what TensorFlow distributed already does for parameter servers).

Given the above, is this the right issue to be commenting on?

**Comment by jmjwozniak** [comment]
Yes, my fork above should address this use case.  It will also let you subdivide the communicators and run Horovods concurrently.

**Comment by alsrgv** [comment]
@ajbouh, are you preferring to pre-allocate workers instead of running multiple smaller MPI jobs because of scheduling inefficiencies?

**Comment by ajbouh** [comment]
Part of it is latency of scheduling. For example, I'd like to be able to
dynamically provision 20+ GPUs in less than 10 seconds.

Part of it is also ease of integration with other tools. For example, being
able to integrate with a REPL means it's also possible to integrate with an
IDE, a notebook, or other development tools. In order to integrate with
these kinds of tools, we can't assume that mpirun can control the lifecycle
of the process that's calling init.

Does this make sense?



On Jun 20, 2018 01:17, "Alex Sergeev" <notifications@github.com> wrote:

@ajbouh <https://github.com/ajbouh>, are you preferring to pre-allocate
workers instead of running multiple smaller MPI jobs because of scheduling
inefficiencies?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/uber/horovod/issues/194#issuecomment-398664627>, or mute
the thread

**Comment by alsrgv** [comment]
@ajbouh, makes sense.  There are few aspects of doing distributed training in REPL-style environment:
1. Distribute the training Python code.
2. Run the training code everywhere.
3. Synchronize gradients during training.

Horovod provides a solution for (3).  (1) and (2) are typically solved with docker or NFS, and `mpirun`.  If you want to do training from the notebook or IDE, you'd need to re-solve (1) and (2).  Depending on a particular solution for that, Horovod integration could be a bit different.

Do you have any ideas for a solution for (1) and (2)?

**Comment by ajbouh** [comment]
Yes, I have a working implementation that solves both (1) and (2) that works with stock TensorFlow. It does this via grpc without NFS or dockerfiles.

**Comment by alsrgv** [comment]
@ajbouh, so you're basically creating a graph on the head process and distributing it to TF workers via gRPC?  You could still use Horovod in this case if you use mpirun to allocate resources, or if you want to be MPI-less you can [allreduce-over-grpc](https://github.com/tensorflow/benchmarks/blob/master/scripts/tf_cnn_benchmarks/allreduce.py) from TF benchmarks.

**Comment by ajbouh** [comment]
Yes, that's what I'm doing essentially.

I hadn't seen allreduce-over-grpc before. What a fantastic pointer. Thank you!

It sounds like the primary value in Horovod from your perspective is tied to using mpirun. Is that right?

**Comment by alsrgv** [comment]
@ajbouh, at this point it is pretty MPI-focused.  We may un-tie it at some point in the future and provide alternative backends, but there are no specific plans yet.

**Comment by ajbouh** [comment]
Makes sense. I'm subscribed to this thread and looking forward to future updates along these lines!

**Comment by alsrgv** [comment]
@jmjwozniak, does #394 solve your need?

**Comment by jmjwozniak** [comment]
Probably- I have not run it yet but the basic idea and changes look very similar to my fork.  However, is it possible to call #394 from C?  That is something my fork supports.

**Comment by ustcyue** [comment]
We are currently implementing a dynamic scheduling feature (number of workers used for a training change dynamically based on the resource availability in a cluster). Most simple way would be restarting the train from a checkpoint when number of workers changed, but this will introduce lots of overhead. If horovod can detach from MPI, it would make this feature much more efficient.

**Comment by alsrgv** [comment]
@ustcyue, detachment from MPI and elastic execution is one of the Horovod projects for 2019.

**Comment by eric-haibin-lin** [comment]
@alsrgv what kind of interface would hvd expose to elastic execution/MPI detachment? Is there a thread for technical discussion or design?

**Comment by alsrgv** [comment]
@eric-haibin-lin, not yet - we will start a discussion thread when thoughts are ideas are fleshed out more concretely.

**Comment by eric-haibin-lin** [comment]
hi @alsrgv just for clarification - does "detachment from MPI" mean horovod is not going to use existing MPI implementations for this feature?

**Comment by alsrgv** [comment]
@eric-haibin-lin, it will be optional.

**Comment by jmjwozniak** [comment]
I am trying to get this working with the new Horovod C API.  Are there any examples of setting up the communicator for Horovod from C?  I am trying it via horovod_init_comm() but it does not seem to have an effect once I run hvd.rank() from Python (I cannot avoid calling hvd.init()).

**Comment by alsrgv** [comment]
@jmjwozniak, can you run `horovod_init_comm` in C before `hvd.init()` and `hvd.rank()` in Python?
